app/assets/javascripts/components/autocomplete.js

```diff
@@ -0,0 +1,221 @@
+        var results = []
+        $options.forEach(function ($el) {
+          results.push({text: $el.textContent, hint: $el.dataset.hint || '', value: $el.value})
+        })
+        syncResults(query
+          ? results.filter(function (result) {
+            var queryLowerCase = query.toLowerCase().trim()
+            var resultTextLowerCase = result.text.toLowerCase()
+            var resultHintLowerCase = result.hint.toLowerCase()
+            var textAndHintLowerCase = resultTextLowerCase + ' ' + resultHintLowerCase
+            var hintAndTextLowerCase = resultHintLowerCase + ' ' + resultTextLowerCase
+
+            var textAndHintContains = textAndHintLowerCase.indexOf(queryLowerCase) !== -1
+            var hintAndTextContains = hintAndTextLowerCase.indexOf(queryLowerCase) !== -1
+
+            return textAndHintContains || hintAndTextContains
+          }) : []
+        )
+      },
+      minLength: 3,
+      autoselect: true,
+      showNoOptionsFound: true,
+      templates: {
+        inputValue: function (result) {
+          return result && result.text
+        },
+        suggestion: function (result) {
+          return result && result.text + '<span class="autocomplete__option-hint">' + result.hint + '</span>'
+        }
+      },
+      onConfirm: function (result) {
+        var value = result && result.value
+        var options = [].filter.call($select.options, function (option) {
+          return option.value === value
+        })
+
+        if (options.length) {
+          options[0].selected = true
+        }
+      }
+    })
+
+    $select.style.display = 'none'
+    $select.id = $select.id + '-select'
+  }
+
+  Autocomplete.prototype.initAutoCompleteInput = function ($input) {
+    var withoutNarrowingResults = this.$module.dataset.autocompleteWithoutNarrowingResults
+
+    var list = document.getElementById($input.getAttribute('list'))
+    var options = []
+
+    if (list) {
+      options = [].map.call(list.querySelectorAll('option'), function (option) {
+        return option.value
+      })
+    }
+
+    if (!options.length) {
+      return
+    }
+
+    new window.accessibleAutocomplete({ // eslint-disable-line no-new, new-cap
+      id: $input.id,
+      name: $input.name,
+      element: this.$module,
+      showAllValues: withoutNarrowingResults,
+      defaultValue: $input.value,
+      autoselect: !withoutNarrowingResults,
+      dropdownArrow: withoutNarrowingResults ? this.dropdownArrow : null,
+      source: function (query, syncResults) {
+        if (withoutNarrowingResults) {
+          syncResults(options)
+        } else {
+          syncResults(query
+            ? options.filter(function (option) {
+              return option.toLowerCase().indexOf(query.toLowerCase()) !== -1
+            }) : []
+          )
+        }
+      }
+    })
+
+    $input.parentNode.removeChild($input)
+  }
+
+  Autocomplete.prototype.initAutoCompleteSearchTopics = function () {
```

Comment:
Is this used?

Reply:
Not by us. But I think it's dangerous to start removing code in case we want to move this out to a separate component in govuk_publishing_apps / design system. That's why I've copied over files as-is.

Convo happening here: https://github.com/alphagov/collections-publisher/pull/664#pullrequestreview-268531804

Reply:
I thought auto-complete was already a component in `govuk-publishing_components`? If you remove the unneeded stuff here, surely it will be easier to see what optional features need to be added to support Content Publisher?

Reply:
As Dilwoar said [here](https://github.com/alphagov/collections-publisher/pull/664#discussion_r309099647) and I said in a commit message, there's an autocomplete component but it doesn't allow you to choose an input value outside of the list of default values (e.g. `10:36am` wouldn't be allowed).

Reply:
On the basis that we're planning to consolidate all Autocomplete component code into one shared component (in govuk_publishing_components) that can then be pulled into content-publisher and collections-publisher...

Our users (content designers) won't care that there's a bit more code in the page, and our developers should never have to look at the code once they realise it's a straightforward copy of whatever is in content-publisher. The only reason we're looking at the HTML/CSS/JS in the first place is that it hasn't been abstracted away in a pattern library we can simply import it from. (There are almost certainly unused features in the other components we're pulling in, but we just don't see the extra code).

When the day comes that this component has been copied to something central, we'll be able to see that the files in Collections Publisher have only been added in one commit, that the commit references the commit in Content Publisher that they were forked from, and therefore we can safely delete the files and import from govuk_publishing_components without worrying that we've lost any functionality.

If, however, we've chopped the code up, we'll have to look at all of the changes and cross-reference them with the version in govuk_publishing_components - creating more work for our present and future selves. In the spirit of optimising code for deletion, I really think putting work into removing bits of code here would be counter-productive.